python/tvm/contrib/pipeline_executor.py

```diff
@@ -501,17 +537,18 @@ class PipelineExecutorFactoryModule(object):
 
     """
 
-    def __init__(self, pipeline_mods, mods_config):
-        mods, config = self.graph_executor_create(pipeline_mods, mods_config)
-        assert (
-            pipeline_executor_enabled()
-        ), "Pipeline executor is not enabled. Please \
-              re-build TVM with USE_PIPELINE_EXECUTOR=ON"
-        pipeline_create = tvm._ffi.get_global_func(
+    def __init__(self, pipeline_libs=None, mods_config=None):
+        self.pipeline_libs = pipeline_libs
+        self.mods_config = mods_config
+        self.pipeline_create = tvm._ffi.get_global_func(
             "tvm.pipeline_executor.create", allow_missing=False
         )
-        assert pipeline_create
-        self.module = pipeline_create(mods, config)
+        self.module = None
+        # Only create pipeline executor when pipeline_libs, mods_config and
+        # self.pipeline_create are not None.
+        if pipeline_libs and mods_config and self.pipeline_create:
```

Comment:
Since you specify that `allow_missing=False`, `self.pipeline_create` should always be non-None.

Reply:
https://github.com/apache/tvm/blob/a1cd6d51b834e7b40b8222f61bc592264845ee39/python/tvm/_ffi/_ctypes/packed_func.py#L292
according the logic , once allow_missing get set into true, get_global_func will return None instead of raise error, so for the said case we may still need to do the None check, please correct me if my finding is wrong.

Reply:
I mean, since at L544 you set `allow_missing=False`, `self.pipeline_create` is always non-None when you reach here.

Reply:
fixed.

Reply:
Following my comment for PipelineModule, I don't think we should allow `pipeline_libs` and `mods_config` to be None:

```python
class PipelineExecutorFactoryModule(object):
  def __init__(self, pipeline_libs, mods_config):
    self.pipeline_libs = pipeline_libs
    self.mods_config = mods_config
    self.pipeline_create = tvm._ffi.get_global_func("tvm.pipeline_executor.create", allow_missing=False)
    graph_executors, config = self.graph_executor_create(pipeline_libs, mods_config)
    self.module = self.pipeline_create(graph_executors, config)

  @staticmethod
  def load_library(config_file_name):
    return PipelineExecutorFactoryModule([], self.pipe_config)
```

So that you can use in PipelineModule:

```python
class PipelineModule(object):
  def __init__(self, module):
    self.module = module.module
    self._get_num_outputs = self.module["get_num_outputs"]

  @staticmethod
  def load_library(config_file_name):
    pipeline_factory = PipelineExecutorFactoryModule.load_library(config_file_name)
    return PipelineModule(pipeline_factory)
```

Reply:
fixed.